test/acquisition/test_knowledge_gradient.py

```diff
@@ -399,6 +400,62 @@ def test_evaluate_q_multi_fidelity_knowledge_gradient(self):
             self.assertTrue(torch.allclose(val, val_exp, atol=1e-4))
             self.assertTrue(torch.equal(qMFKG.extract_candidates(X), X[..., :-n_f, :]))
 
+    def test_evaluate_qMFKG(self):
+        # mock test qMFKG.evaluate() with expand, project & cost aware utility
+        for dtype in (torch.float, torch.double):
+            mean = torch.zeros(1, 1, 1, device=self.device, dtype=dtype)
+            mm = MockModel(MockPosterior(mean=mean))
+            mm._input_batch_shape = torch.Size([1])
+            cau = GenericCostAwareUtility(mock_util)
+            n_f = 4
+            mean = torch.rand(n_f, 2, 1, 1, device=self.device, dtype=dtype)
+            variance = torch.rand(n_f, 2, 1, 1, device=self.device, dtype=dtype)
+            mfm = MockModel(MockPosterior(mean=mean, variance=variance))
+            mfm._input_batch_shape = torch.Size([n_f, 2])
+            with mock.patch.object(MockModel, "fantasize", return_value=mfm) as patch_f:
+                with mock.patch(NO, new_callable=mock.PropertyMock) as mock_num_outputs:
+                    mock_num_outputs.return_value = 1
+                    qMFKG = qMultiFidelityKnowledgeGradient(
+                        model=mm,
+                        num_fantasies=n_f,
+                        X_pending=torch.rand(1, 1, 1, device=self.device, dtype=dtype),
+                        current_value=torch.zeros(1, device=self.device, dtype=dtype),
+                        cost_aware_utility=cau,
+                        project=lambda X: torch.zeros_like(X),
+                        expand=lambda X: torch.ones_like(X),
+                    )
+                    with mock.patch(
+                        "botorch.optim.optimize.optimize_acqf",
+                        return_value=(
+                            torch.ones(1, 1, 1, device=self.device, dtype=dtype),
+                            torch.ones(1, device=self.device, dtype=dtype),
+                        ),
+                    ):
+                        with mock.patch(
+                            "botorch.generation.gen.gen_candidates_scipy",
+                            return_value=(
+                                torch.ones(1, 1, 1, device=self.device, dtype=dtype),
+                                torch.ones(1, device=self.device, dtype=dtype),
+                            ),
+                        ):
+                            val = qMFKG.evaluate(
+                                X=torch.zeros(1, 1, 1, device=self.device, dtype=dtype),
+                                bounds=torch.tensor([[0.0], [1.0]]),
+                                num_restarts=1,
+                                raw_samples=1,
+                            )
```

Comment:
Can use `ExitStack` to get rid of the nested contexts

```suggestion
            with ExitStack() as es:
                patch_f = es.enter_context(
                    mock.patch.object(MockModel, "fantasize", return_value=mfm),
                )
                mock_num_outputs = es.enter_context(
                    mock.patch(NO, new_callable=mock.PropertyMock),
                )
                es.enter_context(
                    mock.patch(
                       "botorch.optim.optimize.optimize_acqf",
                       return_value=(
                           torch.ones(1, 1, 1, device=self.device, dtype=dtype),
                           torch.ones(1, device=self.device, dtype=dtype),
                       ),
                   ),
                )
                es.enter_context(
                    mock.patch(
                        "botorch.generation.gen.gen_candidates_scipy",
                        return_value=(
                            torch.ones(1, 1, 1, device=self.device, dtype=dtype),
                            torch.ones(1, device=self.device, dtype=dtype),
                        ),
                    ),
                )

                mock_num_outputs.return_value = 1
                qMFKG = qMultiFidelityKnowledgeGradient(
                    model=mm,
                    num_fantasies=n_f,
                    X_pending=torch.rand(1, 1, 1, device=self.device, dtype=dtype),
                    current_value=torch.zeros(1, device=self.device, dtype=dtype),
                    cost_aware_utility=cau,
                    project=lambda X: torch.zeros_like(X),
                    expand=lambda X: torch.ones_like(X),
                )
                val = qMFKG.evaluate(
                     X=torch.zeros(1, 1, 1, device=self.device, dtype=dtype),
                     bounds=torch.tensor([[0.0], [1.0]]),
                     num_restarts=1,
                     raw_samples=1,
                 )
```

Reply:
Cool! I've never used `ExitStack` before.